File: app/src/main/java/com/cengels/progressor/units/Convert.java
package com.cengels.progressor.units;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.cengels.progressor.ArrayExtensions;
import com.cengels.progressor.enums.ProgressType;
import com.cengels.progressor.exceptions.IllegalConversionException;

import java.util.IllegalFormatConversionException;

public class Convert {
    @NonNull
    private final UnitValue value;
    @Nullable
    private String[] units;
    private boolean checked = false;

    private Convert(@NonNull UnitValue unitValue) {
        this.value = unitValue;
    }

    @NonNull
    public static Convert from(double value, @NonNull String fromUnit) {
        return new Convert(new UnitValue(value, fromUnit));
    }

    @NonNull
    public static Convert from(@NonNull UnitValue unitValue) {
        return new Convert(unitValue);
    }

    private ProgressType getProgressType() {
        return Unit.getProgressType(this.value.getUnit());
    }

    @Nullable
    private String[] getUnits() {
        if (this.units == null && !this.checked) {
            this.units = Unit.getUnits(this.getProgressType());
            this.checked = true;
        }

        return this.units;
    }

    @NonNull
    public UnitValue getValue() {
        return this.value;
    }

    public UnitValue to(@NonNull String toUnit) {
        if (this.value.getUnit().equals(toUnit)) {
            return this.value;
        }

        if (this.getUnits() == null) {
            throw new IllegalConversionException(this.value.getUnit(), toUnit);
        }

        final String[] units = this.getUnits();

        int index = ArrayExtensions.indexOf(units, this.value.getUnit());
        final int indexTarget = ArrayExtensions.indexOf(units, toUnit);

        if (index == -1 || indexTarget == -1) {
            throw new IllegalConversionException(this.value.getUnit(), toUnit);
        }

        double value = this.value.get();

        if (index < indexTarget) {
            for (int i = index; i < indexTarget; i++) {
                value /= Unit.conversionTable.get(units[i]);
            }
        } else {
            for (int i = index; i > indexTarget; i--) {
                value *= Unit.conversionTable.get(units[i - 1]);
            }
        }

        return new UnitValue(value, toUnit);
    }

    public UnitValue best() {
        if (this.getUnits() == null) {
            return this.value;
        }

        final String[] units = this.getUnits();

        int index = ArrayExtensions.indexOf(units, this.value.getUnit());

        double value = this.value.get();
        String unit = this.value.getUnit();

        while ((index <= units.length - 1 && value < 1 && Unit.conversionTable.containsKey(units[index - 1]))
                || (index >= 0 && Unit.conversionTable.containsKey(unit) && value >= Unit.conversionTable.get(unit))) {
            if (value < 1) {
                value *= Unit.conversionTable.get(units[index - 1]);
                index--;
                unit = units[index];
            } else {
                value /= Unit.conversionTable.get(unit);
                index++;
                unit = units[index];
            }
        }

        return new UnitValue(value, unit);
    }
}

File: app/src/main/kotlin/com/cengels/progressor/units/Convert.kt
package com.cengels.progressor.units

import com.cengels.progressor.enums.ProgressType
import com.cengels.progressor.exceptions.IllegalConversionException

class Convert private constructor(val value: UnitValue) {
    private var units: Array<out String>? = null
    private var checked: Boolean = false

    private val progressType: ProgressType
        get() = Unit.getProgressType(this.value.unit)

    private fun getUnits(): Array<out String>? {
        if (this.units == null && !this.checked) {
            this.units = Unit.getUnits(this.progressType)
            this.checked = true
        }

        return this.units
    }

    /** Converts the [UnitValue] to the specified [unit string][Unit]. */
    fun to(toUnit: String): UnitValue {
        if (this.value.unit == toUnit) {
            return this.value
        }

        val units = this.getUnits() ?: throw IllegalConversionException(this.value.unit, toUnit)

        val index = units.indexOf(this.value.unit)
        val indexTarget = units.indexOf(toUnit)

        if (index == -1 || indexTarget == -1) {
            throw IllegalConversionException(this.value.unit, toUnit)
        }

        var value = this.value.get()

        if (index < indexTarget) {
            for (i in index until indexTarget) {
                value /= Unit.conversionTable.getValue(units[i])
            }
        } else {
            for (i in index downTo indexTarget + 1) {
                value *= Unit.conversionTable.getValue(units[i - 1])
            }
        }

        return UnitValue(value, toUnit)
    }

    /**
     * Gets a new [UnitValue] with the unit that corresponds to the lowest possible number of digits.
     *
     * Example: 1,000m -> 1km
     */
    fun best(): UnitValue {
        val units = this.getUnits() ?: return this.value

        var index = units.indexOf(this.value.unit)

        var value = this.value.get()
        var unit = this.value.unit

        while (index <= units.size - 1 && value < 1 && Unit.conversionTable.containsKey(units[index - 1]) || index >= 0 && Unit.conversionTable.containsKey(unit) && value >= Unit.conversionTable[unit]!!) {
            if (value < 1) {
                value *= Unit.conversionTable.getValue(units[index - 1])
                index--
            } else {
                value /= Unit.conversionTable.getValue(unit)
                index++
            }

            unit = units[index]
        }

        return UnitValue(value, unit)
    }

    companion object {
        /** Prepares a [Convert] builder that converts from the specified [Double] and [unit string][Unit]. */
        fun from(value: Double, fromUnit: String): Convert = Convert(UnitValue(value, fromUnit))

        /** Prepares a [Convert] builder that converts from the specified [UnitValue]. */
        fun from(unitValue: UnitValue): Convert = Convert(unitValue)
    }
}
